Refactors dll injection system to proxy via avifil32.dll.

OmniBlade opened the pull request:
Builds a small stub dll that proxies avifil32.dll and then loads
thyme.dll and runs its Setup_Hooks function. Can in theory support
multiple binaries with different dlls to inject for each.

OmniBlade committed:
Refactors dll injection system to proxy via avifil32.dll.

Builds a small stub dll that proxies avifil32.dll and then loads
thyme.dll and runs its Setup_Hooks function. Can in theory support
multiple binaries with different dlls to inject for each.

xezon commented on src/hookproxy/proxydll.cpp:
```diff
@@ -0,0 +1,251 @@
+    {
+        const char *exe_hash;
+        const char *payload_name;
+    } Payloads[] = {
```
Declaration and instantiation of struct can be split up for better readability.

xezon commented on src/hookproxy/proxydll.cpp:
```diff
@@ -0,0 +1,251 @@
+void Load_Forwarded_Functions()
+{
+    char path[MAX_PATH];
+    CopyMemory(path + GetSystemDirectoryA(path, MAX_PATH - 14), "\\avifil32.dll", 14);
```
Any particular reason to use `CopyMemory` instead of `memcpy` ? Also, can unhardcode value 14 here? Can use `const char dll_name[] = "\\avifil32.dll";` and `sizeof(dll_name)`

xezon commented on src/hookproxy/proxydll.cpp:
```diff
@@ -0,0 +1,251 @@
+{
+    char path[MAX_PATH];
+    CopyMemory(path + GetSystemDirectoryA(path, MAX_PATH - 14), "\\avifil32.dll", 14);
+    HMODULE dll = LoadLibraryA(path);
```
There is a `FreeLibraryA` pairing missing for this?

OmniBlade replied:
The library needs to remain loaded as long as the proxy library is as its used to fulfill the function jumps and this library will only be unloaded when the program closes so its probably safe to let the OS clean up the loaded libraries.

xezon replied:
It's not a deal breaker to not clean up on unload, but it's not very nice either. Next person looking at it may wonder again.

OmniBlade replied:
You aren't supposed to call LoadLibrary from within DLLMain either, this is breaking all kinds of best practice. I will add a note to the detach case in DLLMain that we are deliberately not cleaning up.

xezon commented on src/hookproxy/CMakeLists.txt:
```diff
@@ -0,0 +1,15 @@
+    AVIFileExit=FakeAVIFileExit
+    AVIFileCreateStreamA=FakeAVIFileCreateStreamA
+    AVIStreamSetFormat=FakeAVIStreamSetFormat
+    AVIStreamRelease=FakeAVIStreamRelease
```
Who will call these exports?

OmniBlade replied:
game.dat calls them, or at the very least imports them. They are just jumps to the real dll when called.

xezon commented on src/hookproxy/proxydll.cpp:
```diff
@@ -0,0 +1,251 @@
+        } else if (size == (DWORD)(sizeof(buffer) / sizeof(buffer[0]))) {
+            DWORD size_ = size;
+            do {
+                wchar_t *path_ = (wchar_t *)std::realloc(path, sizeof(wchar_t) * size_ * 2);
```
Instead of reallocating here I suggest to pass sufficient writable buffer as function argument. This way buffer responsibility is handled more elegantly purely by caller. Plus caller can use stack buffer of MAX_PATH. This function overall appears to be quite overengineered. I suggest to remove all unnecessary code and then take a fresh look at it.

OmniBlade replied:
MAX_PATH doesn't really represent the actual maximum path size on windows anymore so its possible for a path to exceed that length up to 32,767 characters according to microsofts docs. I guess we could just allocate that on the stack or as a static buffer from the get go?

xezon replied:
Hmm must be a new thing. I think originally Windows API just handles with MAX_PATH. If you want function to allocate and user to free, you maybe make it obvious in file name and also offer a matching free function for it? It's always odd to return something where user needs to take extra care to free something correctly, whether it be free, delete or delete[] or something else.

OmniBlade replied:
On reflection it probably is overkill as who is likely to be running this and also have the game located in a directory that resolves to greater than MAX_PATH?

xezon commented on src/hookproxy/proxydll.cpp:
```diff
@@ -0,0 +1,251 @@
+    while (true) {
+        DWORD size;
+
+        size = GetModuleFileNameW(nullptr, buffer, sizeof(buffer) / sizeof(buffer[0]));
```
`ARRAY_SIZE(buffer)`

xezon commented on src/hookproxy/proxydll.cpp:
```diff
@@ -0,0 +1,251 @@
+        return false;
+    }
+
+    HMODULE dll = LoadLibraryA(payload);
```
`FreeLibraryA` pair is missing?

xezon commented on src/hookproxy/proxydll.cpp:
```diff
@@ -0,0 +1,251 @@
+ */
+const char* Get_Hash()
+{
+    static char hash[41];
```
`SHA_DIGEST_LENGTH * 2 + 1`

xezon commented on src/hookproxy/proxydll.cpp:
```diff
@@ -0,0 +1,251 @@
+
+    const char *target_hash = Get_Hash();
+
+    for (unsigned i = 0; Payloads[i].exe_hash != nullptr; ++i) {
```
Loop can just condition with `i < ARRAY_SIZE(Payloads)`. There is no need for a null terminated array here.

xezon commented on src/hookproxy/proxydll.cpp:
```diff
@@ -0,0 +1,251 @@
+    CopyMemory(path + GetSystemDirectoryA(path, MAX_PATH - 14), "\\avifil32.dll", 14);
+    HMODULE dll = LoadLibraryA(path);
+    if (dll == nullptr) {
+        MessageBoxA(0, "Cannot load original avifil32.dll library", "Proxy", MB_ICONERROR);
```
Minor: `NULL`

xezon commented on src/hookproxy/proxydll.cpp:
```diff
@@ -0,0 +1,251 @@
+    const char *payload = Get_Payload_Name();
+
+    if (payload == nullptr) {
+        MessageBoxA(0, "Cannot find a payload for this binary", "Proxy", MB_ICONERROR);
```
Minor: `NULL`

xezon commented on src/hookproxy/sha1.cpp:
```diff
@@ -0,0 +1,252 @@
+        std::memcpy(output, &m_finalDigest, SHA_DIGEST_LENGTH);
+    }
+
+    return 20;
```
`SHA_DIGEST_LENGTH` ?

xezon commented on src/hookproxy/sha1.cpp:
```diff
@@ -0,0 +1,252 @@
+ */
+void SHAEngine::Print(const void *buffer, char *output)
+{
+    for (int i = 0; i < 20; i++) {
```
`SHA_DIGEST_LENGTH` ?

xezon commented on src/hookproxy/sha1.cpp:
```diff
@@ -0,0 +1,252 @@
+ */
+int SHAEngine::Print_Result(char *output)
+{
+    uint8_t buffer[20];
```
`SHA_DIGEST_LENGTH` ?

xezon commented on src/hookproxy/sha1.cpp:
```diff
@@ -0,0 +1,252 @@
+void SHAEngine::Print(const void *buffer, char *output)
+{
+    for (int i = 0; i < 20; i++) {
+        std::sprintf(&output[2 * i], "%02x", *(static_cast<const uint8_t *>(buffer) + i));
```
This looks dangerous. What if `output` buffer is not size 2*20? Then we corrupt memory here, no? Any chance to rewrite this to be safe to call always?

xezon commented:
How does game load this? I assume game calls the AVI functions from random dll? But this would mean that you will inject later than some other code and eventually lose the ability to debug very early?

xezon commented on src/hookproxy/proxydll.cpp:
```diff
@@ -0,0 +1,251 @@
+    }
+
+    if (path != buffer) {
+        free(path);
```
shouldn't use `std:free` when using `std::realloc` ?

xezon commented on src/hookproxy/proxydll.cpp:
```diff
@@ -0,0 +1,251 @@
+        free(path);
+    }
+
+    return full_path;
```
Is there a chance function can return non-null but `path` was freed above? If so, then user may double free?

OmniBlade committed:
Removes launcher target.

Refactored dlls now work with a proxydll to auto inject.
Launcher target no longer required.
Updates windows workflow to reflect change.

xezon commented on src/hookproxy/proxydll.cpp:
```diff
@@ -75,96 +76,54 @@ __declspec(naked) void FakeAVIStreamRelease()
  */
 wchar_t *Get_Path()
```
Should return `const wchar_t*` then.

xezon commented on src/hookproxy/proxydll.cpp:
```diff
@@ -0,0 +1,221 @@
+    const char *payload = Get_Payload_Name();
+
+    if (payload == nullptr) {
+        MessageBoxA(nullptr, "Cannot find a payload for this binary", "Proxy", MB_ICONERROR);
```
I suspect most users will not understand this message. Maybe can tell something about game.dat?

xezon commented on src/hookproxy/sha1.cpp:
```diff
@@ -0,0 +1,246 @@
+ */
+const char *SHAEngine::Print_Result()
+{
+    static char output[SHA_DIGEST_LENGTH * 2 + 1];
```
You can make this a class member instead of static.
